File: mobile/app/src/main/java/com/bytecats/metanoia/bible/dao/HighlightsDao.kt
package com.bytecats.metanoia.bible.dao

import android.database.sqlite.SQLiteDatabase

class HighlightsDao(private val openDb: () -> SQLiteDatabase) {

    fun setHighlight(book: String, chapter: Int, verse: Int, color: Int) {
        val db = openDb()
        if (color == 0) {
            db.execSQL(
                "DELETE FROM highlights WHERE book=? AND chapter=? AND verse=?",
                arrayOf(book, chapter.toString(), verse.toString())
            )
        } else {
            db.execSQL(
                "INSERT OR REPLACE INTO highlights (book, chapter, verse, color) VALUES (?, ?, ?, ?)",
                arrayOf(book, chapter.toString(), verse.toString(), color)
            )
        }
        db.close()
    }

    fun getHighlights(book: String, chapter: Int): Map<Int, Int> {
        val map = mutableMapOf<Int, Int>()
        val db = openDb()
        val cursor = db.rawQuery(
            "SELECT verse, color FROM highlights WHERE book=? AND chapter=?",
            arrayOf(book, chapter.toString())
        )
        while (cursor.moveToNext()) map[cursor.getInt(0)] = cursor.getInt(1)
        cursor.close(); db.close()
        return map
    }
}

File: ios/Metanoia/Metanoia/Database/HighlightsDAO.swift
import Foundation
import GRDB

struct HighlightsDAO {
    private let db: DatabaseWriter

    init(db: DatabaseWriter) {
        self.db = db
    }

    func setHighlight(book: String, chapter: Int, verse: Int, color: Int) throws {
        try db.write { db in
            if color == 0 {
                try db.execute(
                    sql: "DELETE FROM highlights WHERE book=? AND chapter=? AND verse=?",
                    arguments: [book, chapter, verse]
                )
            } else {
                try db.execute(
                    sql: "INSERT OR REPLACE INTO highlights (book, chapter, verse, color) VALUES (?, ?, ?, ?)",
                    arguments: [book, chapter, verse, color]
                )
            }
        }
    }

    func getHighlights(book: String, chapter: Int) throws -> [Int: Int] {
        try db.read { db in
            let rows = try Row.fetchAll(
                db,
                sql: "SELECT verse, color FROM highlights WHERE book=? AND chapter=?",
                arguments: [book, chapter]
            )
            var map: [Int: Int] = [:]
            for row in rows {
                map[row["verse"]] = row["color"]
            }
            return map
        }
    }
}
